### ISSUE TITLE
Molecule optimization issues

### ISSUE THREAD

**Issue by Zhanghahah** [zou-group/textgrad#175]
Thanks for your instructions! I notice that you highlight that Textgrad can be applied for molecule optimization, but I can not find any clues in your code repo. Could you please give a specific solution for molecule optimization using Textgrad?  Thanks! :)

**Comment by 13917904019** [comment]
What I don't understand is how new smiles are generated? Are they generated by LLM?

**Comment by tboen1** [comment]
Thanks for your interest in textgrad! Molecule optimization can be framed as a specific application of textgrad's instance optimization pipeline, where the "variable" is a textual representation of a molecule. In our experiments, we choose to encode molecules using the [Simplified Molecular Input Line Entry System (SMILES)](https://en.wikipedia.org/wiki/Simplified_Molecular_Input_Line_Entry_System#:~:text=The%20term%20SMILES%20refers%20to,usually%20apparent%20from%20the%20context.) representation. 

Once we represent a molecule as a SMILES string, we can optimize the SMILES string with respect to any loss computed on the SMILES string using textgrad. For molecular optimization, an implementation of the loss function could be something like a molecular property predictor, and the new SMILES string representing an improved molecule is generated using an LLM in the exact same procedure as in general instance optimization.

Here's a quick example for using textgrad to optimize the druglikeness of molecule as measured by its QED score. A QED score is a scalar between 0 and 1, where 1 indicates better druglikeness. 

```
import textgrad as tg
from rdkit.Chem import MolFromSmiles
from rdkit.Chem.QED import qed

tg.set_backward_engine("gpt-4o")

initial_solution = "CCCCC" # pentane
initial_qed = qed(MolFromSmiles(initial_solution))  # initial qed 0.469

solution = tg.Variable(initial_solution, requires_grad=True, role_description="molecule SMILES string")
qed_score = tg.Variable(str(initial_qed), requires_grad=False, role_description="molecule QED score")

loss_fn = tg.loss.MultiFieldEvaluation(
    evaluation_instruction = tg.Variable(
        "Given a molecule as a SMILES string and it's QED score, evaluate the druglikeness of this molecule.",
        requires_grad=False,
        role_description="molecule evaluation instruction"
    )
    role_descriptions = ["molecule SMILES string", "molecule QED score"]
)

loss = loss_fn(inputs = [solution, qed_score])
print(loss) # "this molecule is not considered very druglike..."

optimizer = tg.TGD(parameters=[solution])

loss.backward()
optimizer.step()

print(solution.value) # CC(C)CC(O)C1=CC=CC=C1
print(qed(MolFromSmiles(solution.value))) # improved qed 0.72
print(solution.gradients) # n-pentane is a simple alkane ... adding hydroxyl, amine ... could enhance drug-likeness."
```

**Comment by 13917904019** [comment]
Thanks for the clear explanation and example—this really helps. Appreciate your work on textgrad!













职庭帆CTO


杭州宠王科技有限公司


***@***.***


13917904019


实践是检验真理的唯一标准。


2025-05-21 星期三







电子名片新出VIP模板啦，快来体验>>




扫一扫,


快速添加名片到手机




--------------------------------------------------------------------------------





------------------ 原始邮件 ------------------




 
 发件人: tboen1  ***@***.***>
 

 
 发送时间: 2025-05-21 07:19:07
 

 
 收件人:zou-group/textgrad ***@***.***> 

 
 抄送:13917904019 ***@***.***>,Comment ***@***.***> 

 
 主题: Re: [zou-group/textgrad] Molecule optimization issues (Issue #175)
 



 
tboen1 left a comment (zou-group/textgrad#175)
 Thanks for your interest in textgrad! Molecule optimization can be framed as a specific application of textgrad's instance optimization pipeline, where the "variable" is a textual representation of a molecule. In our experiments, we choose to encode molecules using the Simplified Molecular Input Line Entry System (SMILES) representation.

 Once we represent a molecule as a SMILES string, we can optimize the SMILES string with respect to any loss computed on the SMILES string using textgrad. For molecular optimization, an implementation of the loss function could be something like a molecular property predictor, and the new SMILES string representing an improved molecule is generated using an LLM in the exact same procedure as in general instance optimization.

 Here's a quick example for using textgrad to optimize the druglikeness of molecule as measured by its QED score. A QED score is a scalar between 0 and 1, where 1 indicates better druglikeness.

 import textgrad as tg
from rdkit.Chem import MolFromSmiles
from rdkit.Chem.QED import qed

tg.set_backward_engine("gpt-4o")

initial_solution = "CCCCC" # pentane
initial_qed = qed(MolFromSmiles(initial_solution)) # initial qed 0.469

solution = tg.Variable(initial_solution, requires_grad=True, role_description="molecule SMILES string")
qed_score = tg.Variable(str(initial_qed), requires_grad=False, role_description="molecule QED score")

loss_fn = tg.loss.MultiFieldEvaluation(
 evaluation_instruction = tg.Variable(
 "Given a molecule as a SMILES string and it's QED score, evaluate the druglikeness of this molecule.",
 requires_grad=False,
 role_description="molecule evaluation instruction"
 )
 role_descriptions = ["molecule SMILES string", "molecule QED score"]
)

loss = loss_fn(inputs = [solution, qed_score])
print(loss) # "this molecule is not considered very druglike..."

optimizer = tg.TGD(parameters=[solution])

loss.backward()
optimizer.step()

print(solution.value) # CC(C)CC(O)C1=CC=CC=C1
print(qed(MolFromSmiles(solution.value))) # improved qed 0.72
print(solution.gradients) # n-pentane is a simple alkane ... adding hydroxyl, amine ... could enhance drug-likeness."

**Comment by tboen1** [comment]
Happy to help! Closing this issue, but feel free to comment if you have any more questions.

**Comment by Zhanghahah** [comment]
Thanks for your response. I suppose that it is a property optimization pipeline supervised by **textgrad** proposed by your work. However, the molecule is generated by a general-purpose LLM (GPT-4, etc), and this part is not in your work's scope, right? If so, we can not ensure the validity of the generated molecules.

**Comment by tboen1** [comment]
Thanks for the question! textgrad is designed as a general-purpose optimization framework, so we did not include any molecule specific modifications to our core optimization procedure. In practice, we find that with a strong base model (e.g. gpt-4o), most generated molecules are valid. 

Alternatively, textgrad does support arbitrary loss functions, which can be constructed to encourage molecule validity. For example, returning to the qed optimization demonstration, a simple modification would be to include a validity penalty in the loss function. 
```
def get_molecule_score(smiles: str) -> str:
    try: 
        mol = MolFromSmiles(initial_solution) # validate with RDKit
        return str(qed(mol))
    except: 
        return "The SMILES string for this molecule is invalid, so the qed could not be computed"

qed_score = tg.Variable(get_molecule_score(smiles_string), requires_grad=False, role_description="molecule QED score")
loss = loss_fn(inputs = [solution, qed_score])
loss.backward()
optimizer.step()
....
```
This is a simple example, but could easily be extended by piping the RDKit errors into molecule score etc.
